dartagnan/src/main/java/com/dat3m/dartagnan/encoding/EncodeSets.java

```diff
@@ -0,0 +1,153 @@
+                        set2.add(t2);
+                    }
+                }
+            }
+        }
+        return map(r1, set1.stream(), r2, set2.stream());
+    }
+
+    @Override
+    public Map<Relation, Stream<Tuple>> visitDomainIdentity(Relation rel, Relation r1) {
+        final RelationAnalysis.Knowledge k1 = ra.getKnowledge(r1);
+        return Map.of(r1, news.stream().flatMap(t -> k1.getMaySet().getByFirst(t.getFirst()).stream().filter(t1 -> !k1.getMustSet().contains(t1))));
+    }
+
+    @Override
+    public Map<Relation, Stream<Tuple>> visitRangeIdentity(Relation rel, Relation r1) {
+        final RelationAnalysis.Knowledge k1 = ra.getKnowledge(r1);
+        return Map.of(r1, news.stream().flatMap(t -> k1.getMaySet().getBySecond(t.getSecond()).stream().filter(t1 -> !k1.getMustSet().contains(t1))));
+    }
+
+    @Override
+    public Map<Relation, Stream<Tuple>> visitInverse(Relation rel, Relation r1) {
+        final RelationAnalysis.Knowledge k1 = ra.getKnowledge(r1);
+        return Map.of(r1, news.stream().map(Tuple::getInverse).filter(t -> k1.getMaySet().contains(t) && !k1.getMustSet().contains(t)));
+    }
+
+    @Override
+    public Map<Relation, Stream<Tuple>> visitTransitiveClosure(Relation rel, Relation r1) {
+        HashSet<Tuple> factors = new HashSet<>();
+        final RelationAnalysis.Knowledge k0 = ra.getKnowledge(rel);
+        for (Tuple t : news) {
+            Event e = t.getSecond();
+            for (Tuple t1 : k0.getMaySet().getByFirst(t.getFirst())) {
+                Tuple t2 = new Tuple(t1.getSecond(), e);
+                if (k0.getMaySet().contains(t2)) {
+                    if (!k0.getMustSet().contains(t1)) {
+                        factors.add(t1);
+                    }
+                    if (!k0.getMustSet().contains(t2)) {
+                        factors.add(t2);
+                    }
+                }
+            }
+        }
+        return map(rel, factors.stream(), r1, lower(news, r1));
+    }
+
+    @Override
+    public Map<Relation, Stream<Tuple>> visitCriticalSections(Relation rscs) {
+        List<Tuple> queue = new ArrayList<>();
+        final RelationAnalysis.Knowledge k0 = ra.getKnowledge(rscs);
+        for (Tuple tuple : news) {
+            Event lock = tuple.getFirst();
+            Event unlock = tuple.getSecond();
+            for (Tuple t : k0.getMaySet().getBySecond(unlock)) {
+                Event e = t.getFirst();
+                if (lock.getCId() < e.getCId() && e.getCId() < unlock.getCId()) {
+                    queue.add(t);
+                }
+            }
+            for (Tuple t : k0.getMaySet().getByFirst(lock)) {
+                Event e = t.getSecond();
+                if (lock.getCId() < e.getCId() && e.getCId() < unlock.getCId()) {
+                    queue.add(t);
+                }
+            }
+        }
+        return Map.of(rscs, queue.stream());
+    }
+
+    private Stream<Tuple> lower(Collection<Tuple> news, Relation relation) {
```

Comment:
I agree. The method returns a stream of the unknwon entries in `news`.
So `extractUnknowns` or `extractUnknownTuples` would be a more descriptive name.
I think it would also be nice to have utility methods in `Knowledge` like `isUnknown` and maybe even `isStatic` (= in must-set) and `isPossible` (= in may-set).

Reply:
I'd prefer `filterUnknowns`.  'Extract' could be read as removing the tuples from `news`.